File: #63_Unique_Paths_II/Solution.java
public class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        if (obstacleGrid == null || obstacleGrid.length == 0) return 0;
        int rows = obstacleGrid.length; int cols = obstacleGrid[0].length;
        if (rows < 1 || cols < 1) return 0;
    	int[] dp = new int[cols];
        if (obstacleGrid[0][0] > 0) return 0;
        dp[0] = 1;
    	for (int i = 0; i < rows; i++) {
    		for (int j = 0; j < cols; j++) {
                if (obstacleGrid[i][j] > 0) {
                    dp[j] = 0;
                }
                else if (j > 0) {
                    dp[j] = dp[j] + dp[j - 1];
                }
    		}
    	}
    	return dp[cols - 1];
    }

    public static void main(String[] args) {
        int[][] t = new int[4][0];
        System.out.println(t.length + ", " + t[0].length);
    }
}

File: #63_Unique_Paths_II/Solution.py
'''
time: O(m * n)
space: O(n)
'''

class Solution:
	# @param {integer[][]} obstacleGrid
	# @return {integer}
	def uniquePathsWithObstacles(self, obstacleGrid):
		if not obstacleGrid or not obstacleGrid[0]:
			return 0
		m = len(obstacleGrid)
		n = len(obstacleGrid[0])
		if obstacleGrid[0][0] == 1:
			return 0
		dp = [0 for j in range(n)]
		k = 0
		while k < n and obstacleGrid[0][k] == 0:
			dp[k] = 1
			k += 1
		for i in range(1, m):
			if obstacleGrid[i][0] == 1:
				dp[0] = 0
			for j in range(1, n):
				dp[j] = dp[j - 1] + dp[j]
				if obstacleGrid[i][j] == 1:
					dp[j] = 0
		return dp[n - 1]

sol = Solution()
print sol.uniquePathsWithObstacles([[0]])